File: src/extras-kcoreaddons/libkshell.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const kshell_enums = enums;
const std = @import("std");

/// ### [Upstream resources](https://api.kde.org/kshell.html)
pub const KShell = extern struct {
    /// ### [Upstream resources](https://api.kde.org/kshell.html)
    ///
    /// The pointer to the underlying Qt C++ object
    ///
    ptr: QtC.KShell,

    pub const _is_KShell = {};

    /// ### DEPRECATED: Use `splitArgs` instead
    ///
    pub const SplitArgs = splitArgs;

    /// ### [Upstream resources](https://api.kde.org/kshell.html#splitArgs)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` cmd: []const u8 `
    ///
    /// ` flags: flag of kshell_enums.Option `
    ///
    /// ` err: *kshell_enums.Errors `
    ///
    pub fn splitArgs(allocator: std.mem.Allocator, cmd: []const u8, flags: i32, err: *i32) []const []const u8 {
        const cmd_str = qtc.libqt_string{
            .len = cmd.len,
            .data = cmd.ptr,
        };
        const _arr: qtc.libqt_list = qtc.KShell_SplitArgs(cmd_str, @bitCast(flags), @ptrCast(err));
        var _str: [*]qtc.libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |i|
                qtc.libqt_string_free(@ptrCast(&_str[i]));
            qtc.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("KShell.splitArgs: Memory allocation failed");
        for (0.._arr.len) |i| {
            const _data_val = _str[i];
            const _buf = allocator.alloc(u8, _data_val.len) catch @panic("KShell.splitArgs: Memory allocation failed");
            @memcpy(_buf, _data_val.data[0.._data_val.len]);
            _ret[i] = _buf;
        }
        return _ret;
    }

    /// ### DEPRECATED: Use `joinArgs` instead
    ///
    pub const JoinArgs = joinArgs;

    /// ### [Upstream resources](https://api.kde.org/kshell.html#joinArgs)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` args: []const []const u8 `
    ///
    pub fn joinArgs(allocator: std.mem.Allocator, args: []const []const u8) []const u8 {
        const args_arr = allocator.alloc(qtc.libqt_string, args.len) catch @panic("KShell.joinArgs: Memory allocation failed");
        defer allocator.free(args_arr);
        for (args, 0..args.len) |str_item, i|
            args_arr[i] = .{
                .len = str_item.len,
                .data = str_item.ptr,
            };
        const args_list = qtc.libqt_list{
            .len = args.len,
            .data = args_arr.ptr,
        };
        var _str = qtc.KShell_JoinArgs(args_list);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("KShell.joinArgs: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `quoteArg` instead
    ///
    pub const QuoteArg = quoteArg;

    /// ### [Upstream resources](https://api.kde.org/kshell.html#quoteArg)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` arg: []const u8 `
    ///
    pub fn quoteArg(allocator: std.mem.Allocator, arg: []const u8) []const u8 {
        const arg_str = qtc.libqt_string{
            .len = arg.len,
            .data = arg.ptr,
        };
        var _str = qtc.KShell_QuoteArg(arg_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("KShell.quoteArg: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `tildeExpand` instead
    ///
    pub const TildeExpand = tildeExpand;

    /// ### [Upstream resources](https://api.kde.org/kshell.html#tildeExpand)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` path: []const u8 `
    ///
    pub fn tildeExpand(allocator: std.mem.Allocator, path: []const u8) []const u8 {
        const path_str = qtc.libqt_string{
            .len = path.len,
            .data = path.ptr,
        };
        var _str = qtc.KShell_TildeExpand(path_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("KShell.tildeExpand: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `tildeCollapse` instead
    ///
    pub const TildeCollapse = tildeCollapse;

    /// ### [Upstream resources](https://api.kde.org/kshell.html#tildeCollapse)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` path: []const u8 `
    ///
    pub fn tildeCollapse(allocator: std.mem.Allocator, path: []const u8) []const u8 {
        const path_str = qtc.libqt_string{
            .len = path.len,
            .data = path.ptr,
        };
        var _str = qtc.KShell_TildeCollapse(path_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("KShell.tildeCollapse: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }
};

/// ### [Upstream resources](https://api.kde.org/kshell.html#public-types)
pub const enums = struct {
    pub const Option = enum(i32) {
        pub const NoOptions: i32 = 0;
        pub const TildeExpand: i32 = 1;
        pub const AbortOnMeta: i32 = 2;
    };

    pub const Errors = enum(i32) {
        pub const NoError: i32 = 0;
        pub const BadQuoting: i32 = 1;
        pub const FoundMeta: i32 = 2;
    };
};

File: src/extras-kcoreaddons/libkshell.cpp
#include <KShell>
#include <QList>
#include <QString>
#include <kshell.h>
#include "libkshell.h"
#include "libkshell.hxx"

libqt_list /* of libqt_string */ KShell_SplitArgs(const libqt_string cmd, int flags, int* err) {
    QString cmd_QString = QString::fromUtf8(cmd.data, cmd.len);
    QList<QString> _ret = KShell::splitArgs(cmd_QString, static_cast<KShell::Options>(flags), reinterpret_cast<KShell::Errors*>(err));
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size())));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        auto _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

libqt_string KShell_JoinArgs(const libqt_list /* of libqt_string */ args) {
    QList<QString> args_QList;
    args_QList.reserve(args.len);
    libqt_string* args_arr = static_cast<libqt_string*>(args.data);
    for (size_t i = 0; i < args.len; ++i) {
        QString args_arr_i_QString = QString::fromUtf8(args_arr[i].data, args_arr[i].len);
        args_QList.push_back(args_arr_i_QString);
    }
    auto _ret = KShell::joinArgs(args_QList);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KShell_QuoteArg(const libqt_string arg) {
    QString arg_QString = QString::fromUtf8(arg.data, arg.len);
    auto _ret = KShell::quoteArg(arg_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KShell_TildeExpand(const libqt_string path) {
    QString path_QString = QString::fromUtf8(path.data, path.len);
    auto _ret = KShell::tildeExpand(path_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KShell_TildeCollapse(const libqt_string path) {
    QString path_QString = QString::fromUtf8(path.data, path.len);
    auto _ret = KShell::tildeCollapse(path_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}
